docs/howtos/ensembling.rst

```diff
@@ -1,307 +1,274 @@
+        apply_fn=cnn.apply, params=params, tx=tx)
   ---
-  @jax.pmap #!
-  def get_initial_params(key):
-    init_val = jnp.ones((1, 28, 28, 1), jnp.float32)
-    initial_params = CNN().init(key, init_val)['params']
-    return initial_params
-
-Note that for the single-model code above, we use `jax.jit`_ to lazily
+  @functools.partial(jax.pmap, static_broadcasted_argnums=(1, 2))  #!
+  def create_train_state(rng, learning_rate, momentum):
+    cnn = CNN()
+    params = cnn.init(rng, jnp.ones([1, 28, 28, 1]))['params']
+    tx = optax.sgd(learning_rate, momentum)
+    return train_state.TrainState.create(
+        apply_fn=cnn.apply, params=params, tx=tx)
+
+Note that for the single-model code above, we use |jax.jit()|_ to lazily
 initialize the model (see `Module.init`_'s documentation for more details).
-For the ensembling case, `jax.pmap`_ will map over the first axis of the
-provided argument ``key`` by default, so we should make sure that we provide
-one key for each device when we call this function later on.
-
-Next we simply do the same for the functions ``create_optimizer``, 
-``train_step``, and ``eval_step``. We also make a minor change to 
-``eval_model``, which ensures the metrics are used correctly in the parallel
-setting.
+For the ensembling case, |jax.pmap()|_ will map over the first axis of the
+provided argument ``rng`` by default, so we should make sure that we provide
+a different value for each device when we call this function later on.
+
+Note also how we specify that ``learning_rate`` and ``momentum`` are static
+arguments, which means the concrete values of these arguments will be used,
+rather than abstract shapes. This is necessary because the provided arguments
+will be scalar values. For more details see `JIT mechanics: tracing and static
+variables`_.
+
+Next we simply do the same for the functions ``apply_model()`` and
+``update_model()``. To compute the predictions from the ensemble, we take the
+average of the individual probabilities. We use |jax.lax.pmean()|_ to compute
+the average *across devices*. This also requires us to specify the
+``axis_name`` to both |jax.pmap()|_ and |jax.lax.pmean()|_.
 
 .. codediff::
   :title_left: Single-model
   :title_right: Ensemble
 
-  # #!
-  def create_optimizer(params, learning_rate=0.1, beta=0.9):
-    optimizer_def = optim.Momentum(learning_rate=learning_rate,
-                                   beta=beta)
-    optimizer = optimizer_def.create(params)
-    return optimizer
-
-  @jax.jit #!
-  def train_step(optimizer, batch):
-    """Train for a single step."""
+  @jax.jit  #!
+  def apply_model(state, images, labels):
     def loss_fn(params):
-      logits = CNN().apply({'params': params}, batch['image'])
-      loss = cross_entropy_loss(logits, batch['label'])
+      logits = CNN().apply({'params': params}, images)
+      one_hot = jax.nn.one_hot(labels, 10)
+      loss = jnp.mean(optax.softmax_cross_entropy(logits=logits, labels=one_hot))
```

Comment:
Nit: You could consider breaking this up in two lines since otherwise the side-by-side is a bit too wide.

Reply:
Done.